packages/babel-parser/src/parser/comments.js

```diff
@@ -284,4 +290,104 @@ export default class CommentsParser extends BaseParser {
 
     stack.push(node);
   }
+
+  processComment(node: Node): void {
+    const { unattachedCommentStack } = this.state;
+    const commentStackLength = unattachedCommentStack.length;
+    if (commentStackLength === 0) return;
+    let i = commentStackLength - 1;
+    const lastComments = unattachedCommentStack[i];
+
+    if (lastComments.start === node.end) {
+      lastComments.leadingNode = node;
+      i--;
+    }
+    // invariant: for all 0 <= j <= i, c = commentStack[j] satisfies c.end < node.end
+    const nodeStart = node.start;
+    for (; i >= 0; i--) {
+      const commentWS = unattachedCommentStack[i];
+      const commentEnd = commentWS.end;
+      if (commentEnd > nodeStart) {
+        // by definition of commentWhiteSpace, this implies commentWS.start > nodeStart
+        // so node can be a containerNode candidate
+        const { containerNode } = commentWS;
+        if (containerNode === null) {
+          commentWS.containerNode = node;
+        } else {
+          if (containerNode.end - containerNode.start < node.end - nodeStart) {
+            unattachedCommentStack.splice(i, 1);
+          } else if (node.type !== "Program") {
+            // we have a node share the same length of containerNode, but its finishNode is invoked later
+            // than containerNode, so this node is the outer node. E.g. ExpressionStatement contains a VariableDeclaration
```

Comment:
Nit: `ExpressionStatement` cannot contain a `VariableDeclaration`, but maybe an `AssignmentExpression`?

Reply:
Yes! You are right. This branch is removed so we are all good here.